### PR TITLE
Add bounds checking for indexed counters, meters, and register arrays

### PR DESCRIPTION
**Pull request by jafingerhut**
with error message printed to log if --log-console or --log-file are
enabled.

Also log messages with possibly useful details on correct accesses to
these stateful objects, for use by P4 programmers when debugging their
code.

### CODE REVIEW

File: targets/simple_switch/primitives.cpp [hunk 1]
```diff
@@ -282,15 +283,42 @@ REGISTER_PRIMITIVE(no_op);
 class execute_meter
   : public ActionPrimitive<MeterArray &, const Data &, Field &> {
   void operator ()(MeterArray &meter_array, const Data &idx, Field &dst) {
-    dst.set(meter_array.execute_meter(get_packet(), idx.get_uint()));
+    unsigned int i = idx.get_uint();
```
Comment (antoninbas) [inline targets/simple_switch/primitives.cpp]: `auto`

File: targets/simple_switch/primitives.cpp [hunk 2]
```diff
@@ -282,15 +283,42 @@ REGISTER_PRIMITIVE(no_op);
...
+    if (i < meter_array.size()) {
+        unsigned int color = meter_array.execute_meter(get_packet(), i);
```
Comment (antoninbas) [inline targets/simple_switch/primitives.cpp]: `auto`

File: targets/simple_switch/primitives.cpp [hunk 3]
```diff
@@ -282,15 +283,42 @@ REGISTER_PRIMITIVE(no_op);
...
+        dst.set(color);
+        BMLOG_TRACE_PKT(get_packet(),
+                        "Updated meter {} at index {},"
```
Comment (antoninbas) [inline targets/simple_switch/primitives.cpp]: Pleas add single quotes around names (`'{}'`)

**Comment by antoninbas** [comment]
Fixes #609

**Comment by jafingerhut** [comment]
When you say "This should probably go inside bm_sim so that all targets can benefit from it", is there a particular file or files you have in mind, that by changing existing methods inside of them will have a similar effect?

Or do you mean that the functions in the file targets/simple_switch/primitives.cpp should be moved to a file inside bm_sim?

Happy to make changes like that if I know what you are looking for.

File: targets/simple_switch/primitives.cpp (hunk 1 above)
Comment (jafingerhut) [reply targets/simple_switch/primitives.cpp]: Done

File: targets/simple_switch/primitives.cpp (hunk 2 above)
Comment (jafingerhut) [reply targets/simple_switch/primitives.cpp]: Done

File: targets/simple_switch/primitives.cpp (hunk 3 above)
Comment (jafingerhut) [reply targets/simple_switch/primitives.cpp]: Done

**Comment by jafingerhut** [comment]
From compiling simple_switch after `./configure 'CXXFLAGS=-O0 -g'`, it gives no error or assert messages when attempting out-of-bounds updates of indexed counters or meters.  It does fail with an assert when attempting to read a register array at an out of bounds index.

I may be drawing the wrong conclusion here, but this seems to suggest that the current code (without this PR's changes) is silently corrupting out-of-bounds memory for such wrong counter and meter updates?  If so, adding a run-time if check would make that check the only one, for counters and meters at least.

**Comment by jafingerhut** [comment]
I have added the requested `#ifndef NDEBUG` around the new out-of-bounds index condition checks and error logging.  Let me know if you see anything else you'd like changed.

**Pull request merged**